File: trees/flatten_tree_linked_list.py
# Java version more intuitive
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
	# @param A : root node of tree
	# @return the root node in the tree
	def flatten(self, A):
		node = root = A
		while root:
			if root.left:
				prev = root.left
				while prev.right:
					prev = prev.right
				prev.right = root.right
				root.right = root.left
				root.left = None
			root = root.right
		return node

"""
Flatten Binary Tree to Linked ListBookmark
Given a binary tree, flatten it to a linked list in-place.

Example :
Given


         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:

   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
Note that the left child of all nodes should be NULL.
"""

File: trees/flatten_tree_linked_list.java
/**
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	public TreeNode flatten(TreeNode A) {
		Queue<TreeNode> queue = new LinkedList<TreeNode>();
		preorder(queue, A);

		if (queue.isEmpty())
			return A;

		TreeNode node = queue.remove();
		TreeNode next;
		A = node;

		while (!queue.isEmpty()) {
			next = queue.remove();
			node.left = null;
			node.right = next;
			node = next;
		}
		return A;
	}


	// Let's try non recursive way
	public void preorder(Queue<TreeNode> queue, TreeNode node) {
		if (node == null)
			return;
		queue.add(node);
		preorder(queue, node.left);
		preorder(queue, node.right);
	}
}